File: Olympiad/COCI/official/2007/contest1_solutions/bond.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 1 - Task BOND
  Programming language Pascal
*)

program bond;

var
   n            : integer;
   i, j, x      : integer;
   vjer         : array[0..20, 0..20] of double;
   vec_rijeseno : array[0..1 shl 21] of boolean;
   memo         : array[0..1 shl 21] of double;

function rijesi(d, s: integer): double;
var
   i        : integer;
   ret, tmp : double;
begin
   if d = n then begin
      rijesi := 1.0;
      exit;
   end;

   if vec_rijeseno[s] then begin
      rijesi := memo[s];
      exit;
   end;

   vec_rijeseno[s] := true;
   ret := 0.0;

   for i:=0 to n-1 do
      if (s and (1 shl i)) = 0 then begin
         tmp := vjer[d, i] * rijesi(d+1, s or (1 shl i));
         if tmp > ret then
            ret := tmp;
      end;

   memo[s] := ret;
   rijesi := ret;
end;

begin
   read(n);
   for i:=0 to n-1 do
      for j:=0 to n-1 do begin
         read(x);
         vjer[i, j] := x / 100;
      end;

   writeln( (rijesi(0, 0)*100):0:6 );
end.

File: Olympiad/COCI/official/2007/contest1_solutions/bond.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 1 - Task BOND
  Programming language C++
*/

#include <cstdio>
using namespace std;

int n;
double prob[21][21];

char vec_rijesio[1<<21];
double memo[1<<21];

double rijesi( int d, int s ) {
   if ( d == n )
      return 1.0;

   if ( vec_rijesio[s] ) return memo[s];
   vec_rijesio[s] = 1;
   double &ret = memo[s];
   ret = 0.0;

   for ( int i=0; i<n; ++i )
      if ( ( s & (1<<i) ) == 0 ) {
         double tmp = prob[d][i] * rijesi(d + 1, s|(1<<i));
         if ( tmp > ret ) ret = tmp;
      }

   return ret;
}

int main() {
   scanf( "%d", &n );

   for ( int i=0; i<n; ++i )
      for ( int j=0; j<n; ++j ) {
         int x;
         scanf( "%d", &x );
         prob[i][j] = x / 100.0;
      }

   double ret = rijesi( 0, 0 );
   printf( "%.6f\n", ret*100.0 );

   return 0;
}
